src/lepton/vp8_decoder.cc

```diff
@@ -380,9 +386,7 @@ namespace{void nop(){}}
 template <class BoolDecoder>
 void VP8ComponentDecoder<BoolDecoder>::reset_all_comm_buffers() {
     for (unsigned int thread_id = 0; thread_id < NUM_THREADS; ++thread_id) {
-        while (!send_to_actual_thread_state.vbuffers[thread_id].empty()) {
-            send_to_actual_thread_state.vbuffers[thread_id].pop();
-        }
+        send_to_actual_thread_state.vbuffers[thread_id].clear();
```

Comment:
Could `send_to_actual_thread_state.free_buffers()` be called here?

Reply:
perhaps, probably :)  

reset_all_comm_buffers() is called twice before starting processing. at these points there should be no raw buffers hold by other objects. i will try! :)

Reply:
Will it be called?

Reply:
As you proposed, `VP8ComponentDecoder<BoolDecoder>::reset_all_comm_buffers()` will only call `VP8ComponentDecoder_SendToVirtualThread::free_buffers()`. Moreover, since they do the same job, I will rename later to `reset_all_comm_buffers()`.

Does it looks like a good plan?

Reply:
At the end of day, I [merged functionality](https://github.com/Bulat-Ziganshin/lepton/commit/de832245798af57ceb90b1ac78b0ed4d757dc48b) of free_buffers() and reset_all_comm_buffers() in all classes having any of this function.

Reply:
Since this update led to bug, and I can't quickly find the reason, I made [limit_memory_with_bug branch](https://github.com/Bulat-Ziganshin/lepton/tree/limit_memory_with_bug) which contains [buggy commit](https://github.com/Bulat-Ziganshin/lepton/commit/de832245798af57ceb90b1ac78b0ed4d757dc48b) and reset this commit back to the initial state in this PR.

On quick glance it seems that we can't do full reset_all_comm_buffers() (which frees allocated_buffers[] and resets vbuffers) at every place where we do free_buffers() (which only frees allocated_buffers[]).

Anyway, answering your initial question - send_to_**actual**_thread_state.free_buffers() function doesn't exist since this object has type VP8ComponentDecoder_SendTo**Actual**Thread, while free_buffers() is function of VP8ComponentDecoder_SendTo**Virtual**Thread.